rllib/models/tests/test_base_model.py

```diff
@@ -0,0 +1,67 @@
+import unittest
+
+from ray.rllib.models.base_model import (
+    SpecDict,
+    UnrollOutputType,
+    TensorDict,
+    Model,
+    RecurrentModel,
+    ForwardOutputType,
+)
+
+
+class NoopRecurrentModelImpl(RecurrentModel):
+    input_spec = output_spec = next_state_spec = prev_state_spec = SpecDict()
+
+    def _initial_state(self):
+        return TensorDict()
+
+    def _unroll(self, inputs: TensorDict, prev_state: TensorDict) -> UnrollOutputType:
+        return (TensorDict(), TensorDict())
+
+
+class NoopModelImpl(Model):
+    input_spec = output_spec = SpecDict()
+
+    def _forward(self, inputs: TensorDict) -> ForwardOutputType:
+        return TensorDict()
+
+
+class TestNoopRecurrentModel(unittest.TestCase):
+    def test_initial_state(self):
+        self.assertEqual(NoopRecurrentModelImpl().initial_state(), TensorDict())
+
+    def test_unroll(self):
+        self.assertEqual(
+            NoopRecurrentModelImpl().unroll(
+                inputs=TensorDict(), prev_state=TensorDict()
+            ),
+            (TensorDict(), TensorDict()),
+        )
+
+    def test_hooks(self):
+        class MyException(Exception):
+            pass
+
+        def exc(a, b):
+            raise MyException()
+
+        m = NoopRecurrentModelImpl()
+        m._check_inputs_and_prev_state = exc
+
+        with self.assertRaises(MyException):
```

Comment:
You can write this inline `self.assertRaises(MyException, lambda m.unroll(TensorDict(), TensorDict()))`.

Reply:
`assertRaises` requires the arguments to be separate from the function:
```
self.assertRaises(MyException, lambda a, b: m.unroll(TensorDict(), TensorDict())), TensorDict({"in": np.arange(3)}), prev_state=TensorDict({"in": np.arange(1)})
```

which I think is ugly. I think the context manager is more readable

Reply:
wait why do you need to pass in parameters? wouldn't this just work: `self.assertRaises(MyException, lambda m.unroll(TensorDict(), TensorDict()))`?

Reply:
I tried that first. `assertRaises` takes a callable and args, you do not actually call the function https://stackoverflow.com/questions/25047256/assertraises-in-python-unit-test-not-catching-the-exception

Reply:
I don't get it you don't need args in the lambda function. look at my unittests for example: https://github.com/ray-project/ray/blob/master/rllib/models/specs/tests/test_specs.py#L94